Python/bltinmodule.c

```diff
@@ -2628,37 +2664,70 @@ builtin_sum_impl(PyObject *module, PyObject *iterable, PyObject *start)
-                    f_result += c;
-                }
-                return PyFloat_FromDouble(f_result);
+                _csum_neumaier_finalize(&f_result);
+                return PyFloat_FromDouble(f_result.sum);
             }
             if (PyFloat_CheckExact(item)) {
-                // Improved Kahan–Babuška algorithm by Arnold Neumaier
-                // Neumaier, A. (1974), Rundungsfehleranalyse einiger Verfahren
-                // zur Summation endlicher Summen.  Z. angew. Math. Mech.,
-                // 54: 39-51. https://doi.org/10.1002/zamm.19740540106
-                // https://en.wikipedia.org/wiki/Kahan_summation_algorithm#Further_enhancements
-                double x = PyFloat_AS_DOUBLE(item);
-                double t = f_result + x;
-                if (fabs(f_result) >= fabs(x)) {
-                    c += (f_result - t) + x;
-                } else {
-                    c += (x - t) + f_result;
+                _csum_neumaier_step(&f_result, PyFloat_AS_DOUBLE(item));
+                _Py_DECREF_SPECIALIZED(item, _PyFloat_ExactDealloc);
+                continue;
+            }
+            if (PyLong_Check(item)) {
+                long value;
+                int overflow;
+                value = PyLong_AsLongAndOverflow(item, &overflow);
+                if (!overflow) {
+                    f_result.sum += (double)value;
+                    Py_DECREF(item);
+                    continue;
                 }
-                f_result = t;
+            }
+            _csum_neumaier_finalize(&f_result);
+            result = PyFloat_FromDouble(f_result.sum);
+            if (result == NULL) {
+                Py_DECREF(item);
+                Py_DECREF(iter);
+                return NULL;
+            }
+            temp = PyNumber_Add(result, item);
+            Py_DECREF(result);
+            Py_DECREF(item);
+            result = temp;
+            if (result == NULL) {
+                Py_DECREF(iter);
+                return NULL;
+            }
+        }
+    }
+
+    if (PyComplex_CheckExact(result)) {
+        Py_complex z = PyComplex_AsCComplex(result);
+        _csum cr_result = {z.real, 0.0};
+        _csum ci_result = {z.imag, 0.0};
+        Py_SETREF(result, NULL);
+        while (result == NULL) {
+            item = PyIter_Next(iter);
+            if (item == NULL) {
+                Py_DECREF(iter);
+                if (PyErr_Occurred()) {
+                    return NULL;
+                }
+                _csum_neumaier_finalize(&cr_result);
+                _csum_neumaier_finalize(&ci_result);
+                return PyComplex_FromDoubles(cr_result.sum, ci_result.sum);
+            }
+            if (PyComplex_CheckExact(item)) {
+                z = PyComplex_AsCComplex(item);
+                _csum_neumaier_step(&cr_result, z.real);
+                _csum_neumaier_step(&ci_result, z.imag);
                 _Py_DECREF_SPECIALIZED(item, _PyFloat_ExactDealloc);
```

Comment:
Does this apply to complex object as well?

It doesn't seem that complex has freelist. Also, there are very few examples of `_PyFloatDealloc` and they are all used on floats.

Unless I am missing something.